### PR TITLE
Simplify custom targets for ninja backend

### PR DESCRIPTION
**Pull request by kinke**
Make custom targets based on 4 predefined rules:
* _custom_in ($in only)
* _custom_out ($out only)
* _custom_in_out ($in, then $out)
* _custom_out_in ($out, then $in)

### CODE REVIEW

**Comment by atilaneves** [comment]
Red CI.

File: payload/reggae/backend/ninja.d [hunk 1]
```diff
@@ -166,179 +166,59 @@ private:
...
-        auto isSameCmd = (in NinjaEntry entry) {
-            bool sameMainLine = entry.mainLine.startsWith(ruleMainLine) &&
-                (entry.mainLine == ruleMainLine || entry.mainLine[ruleMainLine.length] == '_');
-            bool sameCmdLine = entry.paramLines == [ruleCmdLine];
-
-            return sameMainLine && sameCmdLine;
-        };
+        const includeImplicitInputs = (i_out >= 0);
+        // TODO: weird inputsOverride for implicitInput case?!
+        const buildLine = buildLine(target, ruleName, includeImplicitInputs);
```
Comment (kinke) [inline payload/reggae/backend/ninja.d:219]: `implicitInputRule()` used to do some 'weird' stuff wrt. patching the command and preparing some `inputsOverride`, no idea why that was necessary and whether it still is.

I also don't know why the implicit inputs are only included if the custom command has a `$out`.

File: tests/ut/ninja.d [hunk 2]
```diff
@@ -197,24 +173,16 @@ else
...
-                    ["command = gcc $before $out $between $in"]),
-         NinjaEntry("rule translator",
-                    ["command = ./translator $in $out"]),
-         NinjaEntry("rule dmd",
-                    ["command = dmd $before$out $in"])
+        [NinjaEntry(buildPath("build gen/protocol.c gen/protocol.h: _custom_in protocol.proto"),
+                    ["before = ./compiler "]),
+         NinjaEntry(buildPath("build bin/protocol.o: _custom_out  | gen/protocol.c gen/protocol.h"),
+                    ["before = gcc -o ",
+                     "after = $ -c gen/protocol.c"]),
```
Comment (kinke) [inline tests/ut/ninja.d:180]: Previously, the indirect `gen/protocol.c` input seems to have been additionally added as a regular input, and the path replaced as `$in` in the command.

File: tests/ut/ninja.d [hunk 3]
```diff
@@ -197,24 +173,16 @@ else
...
+         NinjaEntry(buildPath("build gen/protocol.d: _custom_out  | gen/protocol.c gen/protocol.h"),
+                    ["before = ./translator gen/protocol.h "]),
```
Comment (kinke) [inline tests/ut/ninja.d:182]: Ditto here, for `gen/protocol.h`.

File: payload/reggae/backend/ninja.d (hunk 1 above)
Comment (atilaneves) [reply payload/reggae/backend/ninja.d:219]: Don't ask me! If the tests pass...

**Review by atilaneves — APPROVED** [review]
I see a bunch of red and the tests pass, so...

**Pull request merged**